examples/hhl.py

```diff
@@ -0,0 +1,311 @@
+in the HHL paper.
+
+This example is an implementation of the HHL algorithm for arbitrary 2x2
+Hermitian matrices. The output of the algorithm is a display of Pauli
+observables of |x>. Note that the accuracy of the result depends on the
+following factors:
+* Register size
+* Choice of parameters C and t
+
+The result is perfect if
+* Each eigenvalue of the matrix is in the form
+
+  2π/t * k/N,
+
+  where 0≤k<N, and N=2^n, where n is the register size. In other words, k is a
+  value that can be represented exactly by the register.
+* C ≤ 2π/t * 1/N, the smallest eigenvalue that can be stored in the circuit.
+
+The result is good if the register size is large enough such that for every
+pair of eigenvalues, the ratio can be approximated by a pair of possible
+register values. Let s be the scaling factor from possible register values to
+eigenvalues. One way to set t is
+
+t = 2π/sN
+
+For arbitrary matrices, because properties of their eigenvalues are typically
+unknown, parameters C and t are fine-tuned based on their condition number.
+
+
+=== REFERENCE ===
+Harrow, Aram W. et al. Quantum algorithm for solving linear systems of
+equations (the HHL paper)
+https://arxiv.org/abs/0811.3171
+
+Coles, Eidenbenz et al. Quantum Algorithm Implementations for Beginners
+https://arxiv.org/abs/1804.03719
+
+=== CIRCUIT ===
+Example of circuit with 2 register qubits.
+
+(0, 0): ─────────────────────────Ry(θ₄)─Ry(θ₁)─Ry(θ₂)─Ry(θ₃)──────────────M──
+                     ┌──────┐    │      │      │      │ ┌───┐
+(1, 0): ─H─@─────────│      │──X─@──────@────X─@──────@─│   │─────────@─H────
+           │         │QFT^-1│    │      │      │      │ │QFT│         │
+(2, 0): ─H─┼─────@───│      │──X─@────X─@────X─@────X─@─│   │─@───────┼─H────
+           │     │   └──────┘                           └───┘ │       │
+(3, 0): ───e^iAt─e^2iAt───────────────────────────────────────e^-2iAt─e^-iAt─
+
+Note: QFT in the above diagram omits swaps, which are included implicitly by
+reversing qubit order for phase kickbacks.
+"""
+
+import math
+import numpy as np
+import cirq
+
+
+class PhaseEstimation(cirq.MultiQubitGate):
+    """
+    A gate for Quantum Phase Estimation.
+
+    unitary is the unitary gate whose phases will be estimated.
+    The last qubit stores the eigenvector; all other qubits store the
+    estimated phase, in big-endian.
+    """
+
+    def __init__(self, num_qubits, unitary):
+        super().__init__(num_qubits)
+        self.U = unitary
+
+    def _decompose_(self, qubits):
+        qubits = list(qubits)
+        yield cirq.H.on_each(*qubits[:-1])
+        yield PhaseKickback(self.num_qubits(), self.U)(*qubits)
+        yield (Qft(self._num_qubits-1)**-1)(*qubits[:-1])
+
+
+class HamiltonianSimulation(cirq.EigenGate, cirq.SingleQubitGate):
+    """
+    A gate that represents e^iAt.
+
+    This EigenGate + np.linalg.eigh() implementation is temporary.
+    Actual implementation will use linear operators.
```

Comment:
Why is the example talking about "temporary" and "actual implementation will"?

Reply:
Not sure what you mean - I can change the wording to something like: "This EigenGate + np.linalg.eigh() implementation is a hack to compute e^iAt; the linear operators framework will be used instead once that's available"

Reply:
I guess I mean that you should do a better job of describing that a non-example implementation on a large matrix can't use this approach, and that it was chosen because it works well for demonstration purposes.

Reply:
I see, I was actually thinking of changing this once linear operators PR is merged, but maybe it's not high in our priorities - it's an example after all. Will update comments.